File: apps/api/app/schemas/product.py
from typing import Any

from pydantic import BaseModel, Field


class EvidenceSnippet(BaseModel):
    source: str
    text: str
    similarity_score: float | None = None
    chunk_type: str | None = None
    aspect: str | None = None
    source_type: str | None = None
    source_platform: str | None = None
    verified_purchase: bool | None = None
    rating: float | None = None
    review_title: str | None = None
    source_url: str | None = None


class CommunityReview(BaseModel):
    id: str
    model: str
    rating: int = Field(ge=1, le=5)
    text: str
    aspect: str
    sentiment: str
    source: str
    timestamp: str


class AspectFrequency(BaseModel):
    aspect: str
    count: int


class CountBreakdownItem(BaseModel):
    label: str
    count: int


class IssueItem(BaseModel):
    title: str
    description: str
    confidence_score: float = Field(ge=0, le=1)


class ReviewSummary(BaseModel):
    overall_sentiment_summary: str
    top_positive_aspects: list[str]
    top_negative_aspects: list[str]
    evidence_snippets: list[EvidenceSnippet]
    chatbot_ready_summary: str


class ProductDecisionSupport(BaseModel):
    recommendation_verdict: str
    quick_verdict: str
    best_for: list[str]
    not_ideal_for: list[str]


class ProductAnalytics(BaseModel):
    review_count: int
    average_rating: float
    positive_percentage: float
    neutral_percentage: float
    negative_percentage: float
    top_issues: list[str]
    top_strengths: list[str]
    issue_frequency: list[AspectFrequency]
    recommendation_score: int = Field(ge=0, le=100)
    source_summary: list[str]
    community_sentiment: str


class ProductBase(BaseModel):
    id: str
    slug: str
    name: str
    family: str
    release_year: int
    chipset: str
    display_type: str
    display_size: str
    battery_summary: str
    camera_summary: str
    storage_options: list[str]
    weight: str
    notable_highlights: list[str]
    short_description: str
    # Intelligence fields — derived from review analytics
    strengths: list[str]
    weaknesses: list[str]
    risk_signals: list[str]
    issue_patterns: list[str]
    community_sentiment: str
    review_count: int
    average_rating: float
    recommendation_score: int = Field(ge=0, le=100)
    target_users: list[str]
    source_summary: list[str]
    user_quotes: list[str]


class ProductSummaryResponse(BaseModel):
    slug: str
    name: str
    review_summary: ReviewSummary


class ProductIssuesResponse(BaseModel):
    slug: str
    name: str
    common_issues: list[IssueItem]
    evidence_snippets: list[EvidenceSnippet]


class ProductDetailResponse(ProductBase):
    review_summary: ReviewSummary
    common_issues: list[IssueItem]
    decision_support: ProductDecisionSupport
    analytics: ProductAnalytics


class ProductListResponse(BaseModel):
    items: list[ProductBase]
    total: int


class ChatRequest(BaseModel):
    slug: str
    question: str
    history: list[dict[str, Any]] = Field(default_factory=list)


class ChatResponse(BaseModel):
    answer: str
    quick_verdict: str
    strengths: list[str]
    weaknesses: list[str]
    issue_warnings: list[str]
    recommendation_score: int = Field(ge=0, le=100)
    evidence_snippets: list[EvidenceSnippet]
    recommendation_verdict: str
    confidence_score: int = Field(ge=0, le=100)
    confidence_label: str | None = None
    evidence_count: int = Field(ge=0)
    top_retrieved_sources: list[str]
    intent_category: str | None = None
    retrieval_debug: list[dict[str, Any]] = Field(default_factory=list)


class ProductReviewsResponse(BaseModel):
    slug: str
    name: str
    total: int
    items: list[CommunityReview]


class NormalizedRealReview(BaseModel):
    id: str
    model: str
    source: str
    rating: float | None = None
    review_text: str
    review_title: str | None = None
    date: str | None = None
    country: str | None = None
    verified_purchase: bool | None = None
    sentiment: str | None = None
    aspect: str | None = None
    source_url: str | None = None


class RealProductReviewsResponse(BaseModel):
    slug: str
    name: str
    total: int
    items: list[NormalizedRealReview]


class RealReviewStatsResponse(BaseModel):
    slug: str
    name: str
    total_reviews: int
    average_rating: float
    positive_count: int
    neutral_count: int
    negative_count: int
    source_distribution: list[CountBreakdownItem]
    rating_distribution: list[CountBreakdownItem]
    verified_review_count: int
    top_aspects: list[AspectFrequency]


class SemanticSearchHit(BaseModel):
    review_id: str
    slug: str | None = None
    model: str
    rating: int | None = Field(default=None, ge=1, le=5)
    text: str
    aspect: str
    sentiment: str
    source: str
    timestamp: str | None = None
    similarity_score: float
    chunk_type: str | None = None
    chunk_label: str | None = None


class SemanticSearchResponse(BaseModel):
    query: str
    total: int
    items: list[SemanticSearchHit]

File: apps/web/types/product.ts
export type EvidenceSnippet = {
  source: string;
  text: string;
  similarity_score?: number | null;
  chunk_type?: string | null;
  aspect?: string | null;
  source_type?: string | null;
  source_platform?: string | null;
  verified_purchase?: boolean | null;
  rating?: number | null;
  review_title?: string | null;
  source_url?: string | null;
};

export type CommunityReview = {
  id: string;
  model: string;
  rating: number;
  text: string;
  aspect: string;
  sentiment: "positive" | "neutral" | "negative";
  source: string;
  timestamp: string;
};

export type AspectFrequency = {
  aspect: string;
  count: number;
};

export type IssueItem = {
  title: string;
  description: string;
  confidence_score: number;
};

export type ReviewSummary = {
  overall_sentiment_summary: string;
  top_positive_aspects: string[];
  top_negative_aspects: string[];
  evidence_snippets: EvidenceSnippet[];
  chatbot_ready_summary: string;
};

export type ProductDecisionSupport = {
  recommendation_verdict: string;
  quick_verdict: string;
  best_for: string[];
  not_ideal_for: string[];
};

export type ProductAnalytics = {
  review_count: number;
  average_rating: number;
  positive_percentage: number;
  neutral_percentage: number;
  negative_percentage: number;
  top_issues: string[];
  top_strengths: string[];
  issue_frequency: AspectFrequency[];
  recommendation_score: number;
  source_summary: string[];
  community_sentiment: string;
};

export type Product = {
  id: string;
  slug: string;
  name: string;
  family: string;
  release_year: number;
  chipset: string;
  display_type: string;
  display_size: string;
  battery_summary: string;
  camera_summary: string;
  storage_options: string[];
  weight: string;
  notable_highlights: string[];
  short_description: string;
  // Intelligence fields derived from review analytics
  strengths: string[];
  weaknesses: string[];
  risk_signals: string[];
  issue_patterns: string[];
  community_sentiment: string;
  review_count: number;
  average_rating: number;
  recommendation_score: number;
  target_users: string[];
  source_summary: string[];
  user_quotes: string[];
};

export type ProductListResponse = {
  items: Product[];
  total: number;
};

export type ProductDetail = Product & {
  review_summary: ReviewSummary;
  common_issues: IssueItem[];
  decision_support: ProductDecisionSupport;
  analytics: ProductAnalytics;
};

export type ProductSummaryResponse = {
  slug: string;
  name: string;
  review_summary: ReviewSummary;
};

export type ProductIssuesResponse = {
  slug: string;
  name: string;
  common_issues: IssueItem[];
  evidence_snippets: EvidenceSnippet[];
};

export type ProductReviewsResponse = {
  slug: string;
  name: string;
  total: number;
  items: CommunityReview[];
};

export type NormalizedRealReview = {
  id: string;
  model: string;
  source: string;
  rating: number | null;
  review_text: string;
  review_title?: string | null;
  date?: string | null;
  country?: string | null;
  verified_purchase?: boolean | null;
  sentiment?: string | null;
  aspect?: string | null;
  source_url?: string | null;
};

export type CountBreakdownItem = {
  label: string;
  count: number;
};

export type RealProductReviewsResponse = {
  slug: string;
  name: string;
  total: number;
  items: NormalizedRealReview[];
};

export type RealReviewStatsResponse = {
  slug: string;
  name: string;
  total_reviews: number;
  average_rating: number;
  positive_count: number;
  neutral_count: number;
  negative_count: number;
  source_distribution: CountBreakdownItem[];
  rating_distribution: CountBreakdownItem[];
  verified_review_count: number;
  top_aspects: AspectFrequency[];
};

export type SemanticSearchHit = {
  review_id: string;
  slug: string | null;
  model: string;
  rating: number | null;
  text: string;
  aspect: string;
  sentiment: "positive" | "neutral" | "negative" | string;
  source: string;
  timestamp: string | null;
  similarity_score: number;
  chunk_type?: string | null;
  chunk_label?: string | null;
};

export type SemanticSearchResponse = {
  query: string;
  total: number;
  items: SemanticSearchHit[];
};

export type ChatRequest = {
  slug: string;
  question: string;
  history?: Array<Record<string, unknown>>;
};

export type ChatResponse = {
  answer: string;
  quick_verdict: string;
  strengths: string[];
  weaknesses: string[];
  issue_warnings: string[];
  recommendation_score: number;
  evidence_snippets: EvidenceSnippet[];
  recommendation_verdict: string;
  confidence_score: number;
  confidence_label?: string | null;
  evidence_count: number;
  top_retrieved_sources: string[];
  intent_category?: string | null;
  retrieval_debug?: Array<Record<string, unknown>>;
};
